x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/enrollment/CreateEnrollmentToken.java

```diff
@@ -0,0 +1,246 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.security.enrollment;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.common.CheckedSupplier;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.UUIDs;
+import org.elasticsearch.common.settings.SecureString;
+import org.elasticsearch.common.ssl.SslUtil;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.json.JsonXContent;
+import org.elasticsearch.core.Tuple;
+import org.elasticsearch.env.Environment;
+import org.elasticsearch.xpack.core.XPackSettings;
+import org.elasticsearch.xpack.core.security.user.ElasticUser;
+import org.elasticsearch.xpack.core.ssl.KeyConfig;
+import org.elasticsearch.xpack.core.ssl.SSLService;
+import org.elasticsearch.xpack.core.ssl.StoreKeyConfig;
+import org.elasticsearch.xpack.security.authc.esnative.ReservedRealm;
+import org.elasticsearch.xpack.security.tool.CommandLineHttpClient;
+import org.elasticsearch.xpack.security.tool.HttpResponse;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.HttpURLConnection;
+import java.net.InetAddress;
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.security.PrivateKey;
+import java.security.cert.X509Certificate;
+import java.util.ArrayList;
+import java.util.Base64;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+public class CreateEnrollmentToken {
+    protected static final String ENROLL_API_KEY_EXPIRATION = "30m";
+
+    private static final Logger logger = LogManager.getLogger(CreateEnrollmentToken.class);
+    private static final String elasticUser = ElasticUser.NAME;
+    private final Environment environment;
+    private final SSLService sslService;
+    private final CommandLineHttpClient client;
+    private final SecureString elasticUserPassword;
+
+    public CreateEnrollmentToken(Environment environment) {
+        this(environment, new CommandLineHttpClient(environment));
+    }
+
+    // protected for testing
+    protected CreateEnrollmentToken(Environment environment, CommandLineHttpClient client) {
+        this.environment = environment;
+        this.sslService = new SSLService(environment);
+        this.client = client;
+        this.elasticUserPassword = ReservedRealm.BOOTSTRAP_ELASTIC_PASSWORD.get(environment.settings());
```

Comment:
This is not what I suggested/asked if we had considered with 

> with user credentials instead of using an origin as one of the internal users ? 

below. See `ClientHelper#executeAsyncWithOrigin` . Furthermore, when we would need to create tokens, the password for the elastic user would have been already set and the bootstrap password won't be valid anymore.